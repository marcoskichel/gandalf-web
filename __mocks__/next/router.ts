const router = { push: jest.fn() }
const useRouter = jest.fn().mockReturnValue(router)

export { useRouter }
