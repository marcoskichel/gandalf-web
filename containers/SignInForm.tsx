import {
  Box,
  Button,
  Container,
  FormControlLabel,
  Grid,
  Link,
  TextField,
  Typography,
} from "@mui/material";
import { FirebaseError } from "firebase-admin";
import { Controller, useForm } from "react-hook-form";
import { useAuth } from "../providers/auth";

interface FormData {
  email: string;
  password: string;
}

const SignInForm = () => {
  const { handleSubmit, reset, control } = useForm<FormData>();
  const { signInWithEmailAndPassword } = useAuth();

  const onSubmit = handleSubmit(async (data) => {
    try {
      await signInWithEmailAndPassword(data.email, data.password);
    } catch (error) {
      const fbError = error as FirebaseError;
      if (
        fbError.code === "auth/user-not-found" ||
        fbError.code === "auth/wrong-password"
      ) {
        alert("invalid user");
      }
    }
  });

  return (
    <Container component="main" maxWidth="xs">
      <Box
        sx={{
          marginTop: 8,
          display: "flex",
          flexDirection: "column",
          alignItems: "center",
        }}
      >
        <Typography component="h1" variant="h5">
          Sign in
        </Typography>
        <Box component="form" onSubmit={onSubmit} noValidate>
          <Controller
            name={"email"}
            control={control}
            render={({ field: { onChange, value } }) => (
              <TextField
                onChange={onChange}
                value={value || ""}
                margin="normal"
                required
                fullWidth
                id="email"
                label="Email Address"
                name="email"
                autoComplete="email"
                autoFocus
              />
            )}
          />
          <Controller
            name={"password"}
            control={control}
            render={({ field: { onChange, value } }) => (
              <TextField
                onChange={onChange}
                value={value || ""}
                margin="normal"
                required
                fullWidth
                name="password"
                label="Password"
                type="password"
                id="password"
                autoComplete="current-password"
              />
            )}
          />
          <Button
            type="submit"
            fullWidth
            variant="contained"
            onClick={handleSubmit(onSubmit)}
            sx={{ mt: 3, mb: 2 }}
          >
            Sign In
          </Button>
          <Grid container>
            <Grid item xs>
              <Link href="#" variant="body2">
                Forgot password?
              </Link>
            </Grid>
            <Grid item>
              <Link href="#" variant="body2">
                {"Don't have an account? Sign Up"}
              </Link>
            </Grid>
          </Grid>
        </Box>
      </Box>
    </Container>
  );
};

export default SignInForm;
