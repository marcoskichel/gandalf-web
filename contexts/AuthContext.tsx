import { auth } from '@config/firebase'
import {
  AuthProvider,
  GoogleAuthProvider,
  User,
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword as firebaseSignInWithEmailAndPassword,
  signOut as firebaseSignOut,
  onIdTokenChanged,
  signInWithPopup,
} from 'firebase/auth'
import nookies from 'nookies'
import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react'

interface AuthContextData {
  loading: boolean
  signUpWithEmailAndPassoword: (
    email: string,
    password: string
  ) => Promise<void>
  signInWithGoogleAccount: () => Promise<void>
  signInWithEmailAndPassword: (email: string, password: string) => Promise<void>
  signOut: () => Promise<void>
  user: User | null
}

const AuthContext = createContext<AuthContextData>({} as AuthContextData)

interface Props {
  children: React.ReactNode
}

const AuthContextProvider = (props: Props) => {
  const { children } = props

  const [loading, setLoading] = useState(false)
  const [user, setUser] = useState<User | null>(null)

  // Update the user state when the user token changes
  useEffect(() => {
    onIdTokenChanged(auth, async (user) => {
      if (user) {
        const token = await user.getIdToken()
        setUser(user)
        nookies.set(undefined, 'token', token, { path: '/' })
      } else {
        setUser(null)
        nookies.set(undefined, 'token', '', { path: '/' })
      }
    })
  }, [])

  // force refresh the token every 10 minutes
  useEffect(() => {
    const handle = setInterval(async () => {
      const user = auth.currentUser
      if (user) {
        await user.getIdToken(true)
      }
    }, 10 * 60 * 1000)

    return () => clearInterval(handle)
  }, [])

  const signUpWithEmailAndPassword = useCallback(
    async (email: string, password: string) => {
      setLoading(true)
      try {
        const credentials = await createUserWithEmailAndPassword(
          auth,
          email,
          password
        )

        setUser(credentials.user)
      } finally {
        setLoading(false)
      }
    },
    []
  )

  const signInWithEmailAndPassword = useCallback(
    async (email: string, password: string) => {
      try {
        setLoading(true)
        const credentials = await firebaseSignInWithEmailAndPassword(
          auth,
          email,
          password
        )

        setUser(credentials.user)
      } finally {
        setLoading(false)
      }
    },
    []
  )

  const signInWithProvider = useCallback(async (provider: AuthProvider) => {
    try {
      setLoading(true)
      const result = await signInWithPopup(auth, provider)
      if (result.user) {
        setUser(result.user)
      }
    } finally {
      setLoading(false)
    }
  }, [])

  const signInWithGoogleAccount = useCallback(async () => {
    const provider = new GoogleAuthProvider()
    return signInWithProvider(provider)
  }, [signInWithProvider])

  const signOut = useCallback(async () => {
    await firebaseSignOut(auth)
    await user?.getIdToken()
  }, [user])

  const value = useMemo(
    () => ({
      loading,
      user,
      signUpWithEmailAndPassoword: signUpWithEmailAndPassword,
      signInWithGoogleAccount,
      signInWithEmailAndPassword,
      signOut,
    }),
    [
      loading,
      user,
      signUpWithEmailAndPassword,
      signInWithGoogleAccount,
      signInWithEmailAndPassword,
      signOut,
    ]
  )

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
}

const useAuth = () => useContext(AuthContext)

export default AuthContext
export { useAuth, AuthContextProvider }
