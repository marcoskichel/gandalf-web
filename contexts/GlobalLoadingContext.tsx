import { createContext, useContext, useMemo, useState } from 'react'

interface ContextProps {
  navigationLoading: boolean
  setNavigationLoading: (value: boolean) => void
}

const GlobalLoadingContext = createContext<ContextProps>({} as ContextProps)

interface ProviderProps {
  children: React.ReactNode
}

const GlobalLoadingContextProvider = (props: ProviderProps) => {
  const { children } = props

  const [navigationLoading, setNavigationLoading] = useState(false)

  const value = useMemo(
    () => ({
      navigationLoading,
      setNavigationLoading,
    }),
    [navigationLoading, setNavigationLoading]
  )

  return (
    <GlobalLoadingContext.Provider value={value}>
      {children}
    </GlobalLoadingContext.Provider>
  )
}

const useGlobalLoading = () => useContext(GlobalLoadingContext)

export default GlobalLoadingContext
export { GlobalLoadingContextProvider, useGlobalLoading }
