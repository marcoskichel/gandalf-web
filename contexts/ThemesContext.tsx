import { Theme, createTheme } from '@mui/material'
import {
  createContext,
  useCallback,
  useContext,
  useMemo,
  useState,
} from 'react'

declare module '@mui/material' {
  interface Palette {
    neutral: Palette['primary']
    default: Palette['primary']
  }
  interface PaletteOptions {
    neutral: PaletteOptions['primary']
    default: PaletteOptions['primary']
  }
}

declare module '@mui/material/Button' {
  interface ButtonPropsColorOverrides {
    neutral: true
    default: true
  }
}

interface AppThemes {
  light: Theme
  dark: Theme
}

const lightTheme = createTheme({
  palette: {
    primary: {
      main: '#6200ea',
      light: '#9d46ff',
      dark: '#320b86',
    },
    neutral: {
      main: '#FFF',
      contrastText: '#616161',
    },
    default: {
      main: '#fff',
      contrastText: '#320b86',
    },
  },
})

const darkTheme = createTheme({
  shape: {
    borderRadius: 8,
  },
  palette: {
    mode: 'dark',
    background: {
      default: '#303030',
      paper: '#424242',
    },
    primary: {
      main: '#6B68F1',
      light: '#D2C1FF',
      dark: '#4A4FD3',
    },
    neutral: {
      main: '#424242',
      contrastText: '#fff',
    },
    default: {
      main: '#303030',
      contrastText: '#fff',
    },
  },
})

interface ContextData {
  currentTheme: Theme
  switchTheme: (name: keyof AppThemes) => void
}

const ThemesContext = createContext<ContextData>({
  currentTheme: lightTheme,
  switchTheme: () => {},
})

const availableThemes = { light: lightTheme, dark: darkTheme }

interface Props {
  children: React.ReactNode
}

// TODO: Implement dark mode
const ThemesContextProvider = (props: Props) => {
  const { children } = props

  const [currentTheme, setCurrentTheme] = useState<Theme>(darkTheme)

  const switchTheme = useCallback((name: keyof AppThemes) => {
    setCurrentTheme(availableThemes[name])
  }, [])

  const value = useMemo(
    () => ({
      currentTheme,
      switchTheme,
    }),
    [currentTheme, switchTheme]
  )

  return (
    <ThemesContext.Provider value={value}>{children}</ThemesContext.Provider>
  )
}

const useThemes = () => useContext(ThemesContext)

export default ThemesContext
export { ThemesContextProvider, useThemes }
