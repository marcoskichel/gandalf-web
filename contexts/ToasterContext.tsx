import { createContext, useContext, useMemo, useState } from 'react'

interface Toast {
  severity?: 'success' | 'error' | 'info' | 'warning'
  message: string
}

interface ToasterContextProps {
  toast?: Toast
  setToast: (toast?: Toast) => void
}

const ToasterContext = createContext<ToasterContextProps>(
  {} as ToasterContextProps
)

interface ToastContextProviderProps {
  children: React.ReactNode
}

const ToasterContextProvider = (props: ToastContextProviderProps) => {
  const { children } = props
  const [toast, setToast] = useState<Toast>()

  const value = useMemo<ToasterContextProps>(
    () => ({
      toast,
      setToast,
    }),
    [toast]
  )

  return (
    <ToasterContext.Provider value={value}>{children}</ToasterContext.Provider>
  )
}

const useToaster = () => useContext(ToasterContext)

export default ToasterContext
export { ToasterContextProvider, useToaster }
