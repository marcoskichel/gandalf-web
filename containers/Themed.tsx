import { ThemeProvider } from "@emotion/react";
import { CssBaseline } from "@mui/material";
import { useThemes } from "../contexts/ThemesContext";

interface Props {
  children: React.ReactNode;
}

const Themed = (props: Props) => {
  const { children } = props;
  const { currentTheme } = useThemes();

  return (
    <ThemeProvider theme={currentTheme}>
      <CssBaseline />
      {children}
    </ThemeProvider>
  );
};

export default Themed;
