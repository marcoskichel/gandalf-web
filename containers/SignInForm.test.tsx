import Routes from '@constants/routes'
import SignInForm from '@containers/SignInForm'
import Themed from '@containers/Themed'
import AuthContext from '@contexts/AuthContext'
import { GlobalLoadingContextProvider } from '@contexts/GlobalLoadingContext'
import { ThemesContextProvider } from '@contexts/ThemesContext'
import ToasterContext from '@contexts/ToasterContext'
import { fireEvent, render, screen } from '@testing-library/react'
import { useRouter } from 'next/router'
import { act } from 'react-dom/test-utils'

jest.mock('firebase/app')
jest.mock('firebase/auth')
jest.mock('firebase/firestore')
jest.mock('next/router')

describe('SignInForm', () => {
  it('renders', () => {
    const { asFragment } = render(
      <GlobalLoadingContextProvider>
        <ThemesContextProvider>
          <Themed>
            <SignInForm />
          </Themed>
        </ThemesContextProvider>
      </GlobalLoadingContextProvider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  describe('Inline validations', () => {
    it('validate email is required', async () => {
      render(
        <GlobalLoadingContextProvider>
          <ThemesContextProvider>
            <Themed>
              <SignInForm />
            </Themed>
          </ThemesContextProvider>
        </GlobalLoadingContextProvider>
      )
      const passwordInput = screen.getByTestId('password-input')
      const btn = screen.getByText('Sign In')

      fireEvent.change(passwordInput, { target: { value: 'password' } })
      await act(() => {
        fireEvent.click(btn)
      })

      const error = screen.getByText('Email Address is a required field')
      expect(error).toBeInTheDocument()
    })

    it('validate email is correctly formatted', async () => {
      render(
        <GlobalLoadingContextProvider>
          <ThemesContextProvider>
            <Themed>
              <SignInForm />
            </Themed>
          </ThemesContextProvider>
        </GlobalLoadingContextProvider>
      )

      const emailInput = screen.getByRole('textbox', { name: 'Email Address' })
      const btn = screen.getByText('Sign In')

      fireEvent.change(emailInput, { target: { value: 'test' } })
      await act(() => {
        fireEvent.click(btn)
      })

      const error = screen.getByText('Must be an email address')
      expect(error).toBeInTheDocument()
    })

    it('validate password is required', async () => {
      render(
        <GlobalLoadingContextProvider>
          <ThemesContextProvider>
            <Themed>
              <SignInForm />
            </Themed>
          </ThemesContextProvider>
        </GlobalLoadingContextProvider>
      )
      const emailInput = screen.getByRole('textbox', { name: 'Email Address' })
      const btn = screen.getByText('Sign In')

      fireEvent.change(emailInput, { target: { value: 'user@test.com' } })
      await act(() => {
        fireEvent.click(btn)
      })

      const error = await screen.findByText('Password is a required field')
      expect(error).toBeInTheDocument()
    })
  })

  describe('Email and Password Authentication', () => {
    const setToast = jest.fn()
    const auth = {
      user: null,
      signOut: jest.fn(),
      loading: false,
      signUpWithEmailAndPassword: jest.fn(),
      signInWithGoogleAccount: jest.fn(),
      signInWithEmailAndPassword: jest.fn(),
    }

    const signIn = async () => {
      render(
        <GlobalLoadingContextProvider>
          <AuthContext.Provider value={auth}>
            <ToasterContext.Provider value={{ setToast }}>
              <ThemesContextProvider>
                <Themed>
                  <SignInForm />
                </Themed>
              </ThemesContextProvider>
            </ToasterContext.Provider>
          </AuthContext.Provider>
        </GlobalLoadingContextProvider>
      )

      const emailInput = screen.getByRole('textbox', { name: 'Email Address' })
      const passwordInput = screen.getByTestId('password-input')
      const btn = screen.getByText('Sign In')

      fireEvent.change(emailInput, { target: { value: 'user@test.com' } })
      fireEvent.change(passwordInput, { target: { value: 'password' } })
      await act(() => {
        fireEvent.click(btn)
      })
    }

    it('shows an error toast when using wrong username', async () => {
      auth.signInWithEmailAndPassword.mockRejectedValueOnce({
        code: 'auth/user-not-found',
      })

      await signIn()
      expect(setToast).toHaveBeenCalledWith({
        message: 'Invalid credentials, please try again.',
        severity: 'error',
      })
    })

    it('shows an error toast when using wrong password', async () => {
      auth.signInWithEmailAndPassword.mockRejectedValueOnce({
        code: 'auth/wrong-password',
      })

      await signIn()
      expect(setToast).toHaveBeenCalledWith({
        message: 'Invalid credentials, please try again.',
        severity: 'error',
      })
    })

    it('redirects authorized users to the Home page', async () => {
      const router = useRouter()
      const mockPush = router.push as jest.Mock

      await signIn()
      expect(mockPush).toHaveBeenCalledWith(Routes.home)
    })
  })
})
