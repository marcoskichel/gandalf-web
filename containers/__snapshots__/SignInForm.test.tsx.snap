// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm renders 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-dvxtzn"
    data-testid="sign-in-form"
  >
    <form
      class="MuiBox-root css-0"
      novalidate=""
    >
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1wb6aqa-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="email"
          id="email-label"
        >
          Email Address
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-125iqkp-MuiFormLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-o15ajz-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="email"
            class="MuiInputBase-input MuiOutlinedInput-input css-p51h6s-MuiInputBase-input-MuiOutlinedInput-input"
            id="email"
            name="email"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-9425fu-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Email Address *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-pcxji5-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Password
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-125iqkp-MuiFormLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-o15ajz-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="current-password"
            class="MuiInputBase-input MuiOutlinedInput-input css-p51h6s-MuiInputBase-input-MuiOutlinedInput-input"
            data-testid="password-input"
            id="password"
            name="password"
            required=""
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-9425fu-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Password *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-19x55hm-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Sign In
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiBox-root css-1ynyhby"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
        >
          OR
        </p>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedNeutral MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedNeutral MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-fnren1-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2724%27%20height=%2724%27/%3e"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="Google Icon"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </span>
        Sign In with Google
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineAlways css-8hyizz-MuiTypography-root-MuiLink-root"
            href="#"
          >
            Forgot password?
          </a>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineAlways css-8hyizz-MuiTypography-root-MuiLink-root"
            href="#"
          >
            Don't have an account? Sign Up
          </a>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
